### PR TITLE
tide: add optional commit message template

### PR DESCRIPTION
**Pull request by diversario**
New Tide config field `merge_commit_template` allows customizing merge commit title and/or body using PullRequest data. The field is optional, as are the title/body keys.

### CODE REVIEW

File: prow/cmd/tide/config.md [hunk 1]
```diff
@@ -17,15 +17,16 @@ The following configuration fields are available:
 * `merge_method`: A key/value pair of an `org/repo` as the key and merge method to override
    the default method of merge as value. Valid options are `squash`, `rebase`, and `merge`.
    Defaults to `merge`.
+* `merge_commit_template`: A key/value pair of an `org/repo` as the key and a map `title: "Go template"`, `body: "Go template"` as value. A `PullRequest` is provided as input to the template (see `PullRequest` type in [tide.go](/prow/cmd/tide.go)). This field and map keys are optional.
```
Comment (stevekuznetsov) [inline prow/cmd/tide/config.md]: This is a bit hard to follow - maybe:

```
 * `merge_commit_template`: A mapping from `org/repo` to a set of Go templates to use when creating the title and body of merge commits. Go templates are evaluated with a `PullRequest`  (see `PullRequest` type in [tide.go](/prow/cmd/tide.go)). This field and map keys are optional.
```

File: prow/config/tide.go [hunk 2]
```diff
@@ -81,6 +91,11 @@ type Tide struct {
 	// the default method of merge. Valid options are squash, rebase, and merge.
 	MergeType map[string]github.PullRequestMergeType `json:"merge_method,omitempty"`
 
+	// A key/value pair of an org/repo as the key and Go template to override
+	// the default merge commit title and/or message. Template is passed the
+	// PullRequest struct (prow/github/types.go#PullRequest)
+	MergeTemplate map[string]*TideMergeCommitTemplate `json:"merge_commit_template,omitempty"`
```
Comment (stevekuznetsov) [inline prow/config/tide.go]: Can we map from the org/repo not to a pointer? In what case should someone want to map to a `nil`?

File: prow/config/tide.go [hunk 3]
```diff
@@ -140,6 +155,22 @@ func (t *Tide) MergeMethod(org, repo string) github.PullRequestMergeType {
 	return v
 }
 
+// MergeCommitTemplate returns the merge method to use for a repo, or an empty string
```
Comment (stevekuznetsov) [inline prow/config/tide.go]: Godoc typo -- returns template, not method

Comment (stevekuznetsov) [reply prow/config/tide.go]: And never an empty string :)

File: prow/cmd/tide/config.md (hunk 1 above)
Comment (diversario) [reply prow/cmd/tide/config.md]: Haha, thank you! The docs are the second hardest thing after variable names 😅

File: prow/config/tide.go (hunk 2 above)
Comment (diversario) [reply prow/config/tide.go]: So, initially I did
```
MergeTemplate map[string]TideMergeCommitTemplate `json:"merge_commit_template,omitempty"`
```
And then when I was writing tests [this bit here](https://github.com/diversario/test-infra/blob/tide-allow-commit-msg-customization/prow/config/config.go#L1032) blew up with [this error](https://stackoverflow.com/questions/32751537/why-do-i-get-a-cannot-assign-error-when-setting-value-to-a-struct-as-a-value-i). Googling led me to a suggestion to use a pointer to the struct. Then I was able to write the tests.

As for mapping to a `nil`, I think config like
```yaml
tide:
  merge_commit_template:
```
would result in the pointer being `nil` (and we check for that so 🤞 no 💥).

I definitely do agree about less pointers == cleaner/simpler code. If there's a better way to do this thing I'm all 👂 !

Comment (stevekuznetsov) [reply prow/config/tide.go]: Ah, I see -- I think you can just do this:


```go
for name, templates := range c.Tide.MergeTemplate {
	if templates != nil {
		if templates.TitleTemplate != "" {
			titleTemplate, err := template.New("CommitTitle").Parse(templates.TitleTemplate)
			if err != nil {
				return fmt.Errorf("parsing template for commit title: %v", err)
			}
			templates.Title = titleTemplate
			c.Tide.MergeTemplate[name] = templates
		}
		if templates.BodyTemplate != "" {
			bodyTemplate, err := template.New("CommitBody").Parse(templates.BodyTemplate)
			if err != nil {
				return fmt.Errorf("parsing template for commit body: %v", err)
			}
			templates.Body = bodyTemplate
			c.Tide.MergeTemplate[name] = templates
		}
	}
}
```

Comment (diversario) [reply prow/config/tide.go]: I _think_ that's what I did (that's how I ran into the error). But, I will double-check!

Comment (diversario) [reply prow/config/tide.go]: Turns out I was doing
```
c.Tide.MergeTemplate[name].Value = titleTemplate
```
stuff.

**Comment by fejta** [comment]
/uncc
Cole want to review this? Assign back to me if not

File: prow/cmd/tide/config.md [hunk 4]
```diff
@@ -17,15 +17,16 @@ The following configuration fields are available:
 * `merge_method`: A key/value pair of an `org/repo` as the key and merge method to override
    the default method of merge as value. Valid options are `squash`, `rebase`, and `merge`.
    Defaults to `merge`.
+* `merge_commit_template`: A mapping from `org/repo` to a set of Go templates to use when creating the title and body of merge commits. Go templates are evaluated with a `PullRequest`  (see `PullRequest` type in [tide.go](/prow/cmd/tide.go)). This field and map keys are optional.
```
Comment (cjwagner) [inline prow/cmd/tide/config.md]: Please add a link to https://godoc.org/k8s.io/test-infra/prow/tide#PullRequest

Comment (cjwagner) [inline prow/cmd/tide/config.md]: Please clarify that the map key can be `org` or `org/repo`.

File: prow/config/tide.go [hunk 5]
```diff
@@ -140,6 +155,22 @@ func (t *Tide) MergeMethod(org, repo string) github.PullRequestMergeType {
...
+func (t *Tide) MergeCommitTemplate(org, repo string) TideMergeCommitTemplate {
+	name := org + "/" + repo
+
+	v, ok := t.MergeTemplate[name]
+	if !ok {
+		if ov, found := t.MergeTemplate[org]; found {
+			return ov
+		}
+
+		return TideMergeCommitTemplate{}
```
Comment (cjwagner) [inline prow/config/tide.go]: nit: The conditional block could be simplified to:
```go
return t.MergeTemplate[org]
```

File: prow/config/tide_test.go [hunk 6]
```diff
@@ -148,6 +148,80 @@ func TestMergeMethod(t *testing.T) {
...
+	var testcases = []struct {
+		org      string
+		repo     string
+		expected TideMergeCommitTemplate
+	}{
+		{
+			"kubernetes",
+			"kubernetes",
+			TideMergeCommitTemplate{},
+		},
```
Comment (cjwagner) [inline prow/config/tide_test.go]: nit: Please specify the field names for these struct literals.

File: prow/tide/tide_test.go [hunk 7]
```diff
@@ -25,6 +25,7 @@ import (
 	"net/http/httptest"
 	"reflect"
 	"testing"
+	textTemplate "text/template"
```
Comment (cjwagner) [inline prow/tide/tide_test.go]: nit: Is this necessary?

File: prow/config/tide_test.go (hunk 6 above)
Comment (diversario) [reply prow/config/tide_test.go]: Does this mean
```
TideMergeCommitTemplate{}
```
is equivalent to
```
TideMergeCommitTemplate{
  Title: nil,
  Body: nil,
}
```
? I guess fields in the struct are set to their default values?

File: prow/config/tide.go (hunk 5 above)
Comment (diversario) [reply prow/config/tide.go]: Could you please explain why this works?

Comment (stevekuznetsov) [reply prow/config/tide.go]: When you index into a map with a key that does not exist in it:

```go
item, exists: myMap["somekey"]
```

`if !exists` will mean that `item` is the zero value for the type, in this case `TideMergeCommitTemplate{}`

File: prow/config/tide_test.go [hunk 8]
```diff
@@ -148,6 +148,83 @@ func TestMergeMethod(t *testing.T) {
...
+
+	var testcases = []struct {
+		org      string
+		repo     string
+		expected TideMergeCommitTemplate
+	}{
+		{
+			"kubernetes",
+			"kubernetes",
+			TideMergeCommitTemplate{
```
Comment (stevekuznetsov) [inline prow/config/tide_test.go]: The intent was :-1: 

```go
{
    org: "kubernetes",
    repo: "kubernetes",
    expected: ...
}
```

If a field is omitted it is not necessary to name it and set it to `nil` but if if a field is set we want to use literal names for the fields so ordering is irrelevant and refactoring it is easier in the future. @cjwagner weird that `lint` didn't catch this with `go vet`

**Comment by stevekuznetsov** [comment]
/ok-to-test
@diversario one last comment!

**Comment by fejta** [comment]
/retest
ref https://github.com/kubernetes/test-infra/pull/12418

File: prow/config/tide_test.go (hunk 8 above)
Comment (diversario) [reply prow/config/tide_test.go]: Gotcha, I was following the structure from the test above it.
